File: src/main/java/com/mmust/evaluators/impl/ShuntingYardMathEvaluator.java
package com.mmust.evaluators.impl;

import com.mmust.evaluators.InfixToPostfixParser;
import com.mmust.evaluators.MathEvaluator;
import com.mmust.evaluators.PostfixEvaluator;
import com.mmust.token.Token;
import com.mmust.tokenizer.Tokenizer;

public class ShuntingYardMathEvaluator implements MathEvaluator {

	private InfixToPostfixParser infixToPostfixParser = new ShuntingYardInfixToPostfixParser();
	private PostfixEvaluator postfixEvaluator = new StackPostfixMathExpressionEvaluator();

	@Override
	public double eval(String infixExpression) {
		Tokenizer tokenizer = new Tokenizer(infixExpression);
		Iterable<Token> postfixExpression = infixToPostfixParser.parse(tokenizer);
		return postfixEvaluator.eval(postfixExpression);
	}
}

File: src/main/kotlin/com/mmust/evaluators/impl/ShuntingYardMathEvaluator.kt
package com.mmust.evaluators.impl

import com.mmust.evaluators.MathEvaluator
import com.mmust.tokenizer.Tokenizer

class ShuntingYardMathEvaluator : MathEvaluator {

    private val infixToPostfixParser = ShuntingYardInfixToPostfixParser()
    private val postfixEvaluator = StackPostfixMathExpressionEvaluator()

    override fun eval(infixExpression: String): Double {
        val tokens = Tokenizer(infixExpression).parse()
        val postfixExpression = infixToPostfixParser.parse(tokens)
        return postfixEvaluator.eval(postfixExpression)
    }

    override fun invoke(infixExpression: String): Double {
        return eval(infixExpression)
    }
}
